File: haskell/src/Solutions/Euler001.hs
module Solutions.Euler001 where

-- k + 2k + ... + k (bound `div` k) = k (1 + ... + bound `div` k)
sumDivisibleUpTo :: Int -> Int -> Int
sumDivisibleUpTo k bound = let l = bound `div` k in k * l * (l + 1) `div` 2

euler001 :: IO String
euler001 = return $ show $ sumDivisibleUpTo 3 999 + sumDivisibleUpTo 5 999 - sumDivisibleUpTo 15 999

File: rust/src/solutions/euler001.rs
fn sum_divisible_up_to(k: i32, bound: i32) -> i32 {
    // k + 2k + ... + k (bound / k) = k (1 + ... + bound / k)
    let l = bound / k;
    k * l * (l + 1) / 2
}

pub fn euler001() -> String {
    (sum_divisible_up_to(3, 999) + sum_divisible_up_to(5, 999) - sum_divisible_up_to(15, 999))
        .to_string()
}
